### PR TITLE
MEN-6316: Implement all the non-Download states + support commits

### PR DESCRIPTION
**Pull request by kacf**

### CODE REVIEW

File: mender-update/update_module/v3/platform/update_module_call/update_module_call.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+namespace update {
+namespace update_module {
+namespace v3 {
+
+namespace error = mender::common::error;
+namespace procs = mender::common::processes;
+namespace fs = boost::filesystem;
+
+ExpectedExitStatus CallState(
+	const std::string process, State state, const string directory, string &procOut) {
```
Comment (oleorhagen) [inline mender-update/update_module/v3/platform/update_module_call/update_module_call.cpp]: string refs?

File: mender-update/update_module/v3/update_module.cpp [hunk 2]
```diff
@@ -43,6 +44,18 @@ UpdateModule::UpdateModule(
 		path::Join(conf::paths::DefaultModulesWorkPath, "payloads", "0000", "tree");
 }
 
+error::Error UpdateModule::CallStateNoOutput(State state) {
+	std::string processStdOut;
```
Comment (oleorhagen) [inline mender-update/update_module/v3/update_module.cpp]: Q: Why do we need to collect the output, even if temporary?

File: mender-update/update_module/v3/update_module.cpp [hunk 3]
```diff
@@ -52,47 +65,87 @@ error::Error UpdateModule::Download() {
...
+	auto procExit =
+		CallState(GetModulePath(), State::NeedsReboot, GetModulesWorkPath(), processStdOut);
+	if (!procExit) {
+		return expected::unexpected(procExit.error());
+	} else if (procExit.value() != 0) {
+		return expected::unexpected(error::MakeError(
+			error::GenericError, "Process exited with error: " + std::to_string(procExit.value())));
+	}
+	if (0 == processStdOut.find("Yes")) {
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply mender-update/update_module/v3/update_module.cpp]: Also, should this not be `string::npos` if it's not found? :thinking:

File: mender-update/update_module/v3/update_module.cpp [hunk 4]
```diff
@@ -52,47 +65,87 @@ error::Error UpdateModule::Download() {
...
+	auto procExit =
+		CallState(GetModulePath(), State::SupportsRollback, GetModulesWorkPath(), processStdOut);
+	if (!procExit) {
+		return expected::unexpected(procExit.error());
+	} else if (procExit.value() != 0) {
+		return expected::unexpected(error::MakeError(
+			error::GenericError, "Process exited with error: " + std::to_string(procExit.value())));
+	}
+	if (0 == processStdOut.find("Yes")) {
+		return expected::ExpectedBool(true);
```
Comment (oleorhagen) [inline mender-update/update_module/v3/update_module.cpp:95]: No need to explicitly cast here

File: mender-update/update_module/v3/update_module_call.hpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <mender-update/update_module/v3/update_module.hpp>
+
+
+namespace mender {
+namespace update {
+namespace update_module {
+namespace v3 {
+
+using ExpectedExitStatus = expected::expected<int, error::Error>;
```
Comment (oleorhagen) [inline mender-update/update_module/v3/update_module_call.hpp]: Just an alias for `ExpectedInt`, no?

File: mender-update/update_module/v3/update_module_call.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+ExpectedExitStatus CallState(
+	const string process, State state, const string directory, string &procOut);
```
Comment (oleorhagen) [inline mender-update/update_module/v3/update_module_call.hpp]: Q: Const?

File: mender-update/update_module/v3/platform/update_module_call/update_module_call.cpp [hunk 7]
```diff
@@ -30,7 +30,7 @@ namespace procs = mender::common::processes;
 namespace fs = boost::filesystem;
 
 ExpectedExitStatus CallState(
-	const std::string process, State state, const string directory, string &procOut) {
+	const string &process, State state, const string &directory, string &procOut) {
```
Comment (oleorhagen) [inline mender-update/update_module/v3/platform/update_module_call/update_module_call.cpp]: Just squash this with the previous commit?

File: common/error.hpp [hunk 8]
```diff
@@ -52,6 +52,8 @@ class Error {
 	std::error_condition code;
 	std::string message;
 
+	Error() {
```
Comment (oleorhagen) [inline common/error.hpp:55]: :+1:

File: mender-update/update_module/v3/update_module.cpp (hunk 4 above)
Comment (kacf) [reply mender-update/update_module/v3/update_module.cpp:95]: Good point, fixed!

File: mender-update/update_module/v3/update_module.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kacf) [reply mender-update/update_module/v3/update_module.cpp]: > Still, all our other stuff does it the other way around, would you mind inverting it?

Done! After my squashing, only one such case remained, that I could see, at least.

**Comment by kacf** [comment]
Updated!

**Pull request merged**